### DISCUSSION TITLE
Motif and discord discovery

### DISCUSSION THREAD

**Discussion by rajvardhan21** [TDAmeritrade/stumpy#353] [General]
Hello,

In stemgen dataset,m=640 was selected on the basis of domain knowledge,but why 643 & 8724 was selected for motif ,how did you deduce that?![Screenshot_20210301-011655__01.jpg](https://user-images.githubusercontent.com/77498003/109431379-47ff5e80-7a2c-11eb-8e77-ffefb0abbbc8.jpg)

**Comment by rajvardhan21** [comment]
> Hello,
> In stemgen dataset,m=640 was selected on the basis of domain knowledge,but why 643 & 8724 was selected for motif ,how did you deduce that?

Also, How can we print matrix profile![Screenshot_20210228-023035__01.jpg](https://user-images.githubusercontent.com/77498003/109431448-b80de480-7a2c-11eb-9515-fab1d11fc228.jpg)

**Reply by seanlaw** [reply]
@rajvardhan21 In your original post, there are two plots that are shown. The top plot is the "Steam Flow" and corresponds to the raw data stored in `steam_df['steam flow']`. The bottom plot is the "Matrix Profile" and corresponds to the first column of `mp` (i.e., `mp[:, 0]`). The `mp` was computed in cell number 5:

```
m = 640
mp = stumpy.stump(steam_df['steam flow'], m)
```

And it is clearly stated in the tutorial that `mp`:

> is an array that contains all of the matrix profile values (i.e., z-normalized Euclidean distance to your nearest neighbor) and matrix profile indices in the first and second columns, respectively

So, the first column of `mp` contains the matrix profile values and you can print them in the notebook with:

```
print(mp[:, 0])
```

The location of the motif pair is obtained by finding the index location of the two smallest values in the matrix profile distance. 

```
print(np.argsort(mp[:, 0])[:2])
```

The discord (a single subsequence) can be found via:

```
print(np.argsort(mp[:, 0])[:-1])
```

> Also, How can we print matrix profile

To be absolutely clear, the matrix profile is a 1-dimensional vector (or list) of values. The image that you have pasted is actually the "distance matrix" and is not available from STUMPY.

**Reply by seanlaw** [reply]
Additionally, I think you've helped us identify an area of our tutorial that can be improved so thank you!

**Reply by seanlaw** [reply]
@rajvardhan21 I've updated the tutorial and I hope it is clearer now:

https://stumpy.readthedocs.io/en/latest/Tutorial_STUMPY_Basics.html#Find-a-Motif-Using-STUMP

**Reply by rajvardhan21** [reply]
Hi Sean,

thanks for your reply, can you help me with this error. this error shows up while finding out discord

![image](https://user-images.githubusercontent.com/77498003/109490946-f5b35180-7aae-11eb-9d04-0af067e9fe17.png)

**Reply by seanlaw** [reply]
@rajvardhan21 It looks like you are missing a closing parenthesis:

```
print(
    np.argsort(mp[:, 0])[-1]
)
``` 

Also, notice that the tutorial uses `[-1]` and not `[:-1]` (note that there is no colon).

**Reply by rajvardhan21** [reply]
Thanks sean,

1. can you help me visualize the mp array data in matrix form like you have shown in the presentation? 
2. do you have any used case for finding the best value of m

**Reply by seanlaw** [reply]
> 1. can you help me visualize the mp array data in matrix form like you have shown in the presentation?

As I had mentioned above, it is not possible to output the distance matrix (only the matrix profile) in STUMPY. This is simply not feasible for longer time series as you would run out of computer memory. 

> 2. do you have any used case for finding the best value of m

For finding `m`, this is unique/different for every time series and so there is no easy answer aside from trial and error. I recommend taking a look at [this paper](https://www.cs.ucr.edu/~eamonn/PAN_SKIMP%20%28Matrix%20Profile%20XX%29.pdf) for ideas.

**Reply by rajvardhan21** [reply]
Thanks sean,

can we save mp array data and use mp.describe function?

also, can we plot next closest values (3-4 values) of mp on graph, if yes how?
